### PR TITLE
feat: use online dist info

### PR DESCRIPTION
**Pull request by hacdias**
Fetches the information of available versions, platforms and archs directly from the website.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "go-ipfs-dep",
-  "version": "0.4.18",
+  "version": "0.4.19",
```
Comment (alanshaw) [inline package.json]: This will be done as part of the release process.

File: src/index.js [hunk 2]
```diff
@@ -25,94 +25,112 @@ const path = require('path')
...
+  const distUrl = (goIpfsInfo && goIpfsInfo.distUrl)
+    ? pkg['go-ipfs'].distUrl
+    : 'https://dist.ipfs.io'
+
+  return {
+    version: process.env.TARGET_VERSION || version || goIpfsVersion,
+    platform: process.env.TARGET_OS || platform || goenv.GOOS,
+    arch: process.env.TARGET_ARCH || arch || goenv.GOARCH,
+    distUrl: process.env.GO_IPFS_DIST_URL || distUrl,
+    installPath: installPath ? path.resolve(installPath) : path.resolve(process.cwd())
```
Comment (alanshaw) [inline src/index.js]: No need to resolve `process.cwd()`

File: src/index.js [hunk 3]
```diff
@@ -25,94 +25,112 @@ const path = require('path')
...
-
-        unpack(res)
-      })
   })
 }
 
-module.exports = download
+async function ensureVersion ({ version, distUrl }) {
+  const res = await get(`${distUrl}/go-ipfs/versions`)
+  const versions = res.split('\n').filter(el => el !== '')
```
Comment (alanshaw) [inline src/index.js]:
```suggestion
  const versions = res.trim().split('\n')
```

File: src/index.js [hunk 4]
```diff
@@ -25,94 +25,112 @@ const path = require('path')
...
+}
 
-    // Success callback wrapper
-    // go-ipfs contents are in 'go-ipfs/', so append that to the path
-    const done = () => resolve({
-      fileName: fileName,
-      installPath: path.join(installPath, '/go-ipfs/')
+function download ({ installPath, url }) {
+  return new Promise((resolve, reject) => {
+    request.get(url, (err, res, body) => {
```
Comment (alanshaw) [inline src/index.js]: Might be easier to use `node-fetch` instead.

File: src/index.js [hunk 5]
```diff
@@ -25,94 +25,112 @@ const path = require('path')
...
+  const args = cleanArguments(...arguments)
+  const url = await getURL(args)
+
+  process.stdout.write(`Downloading ${url}\n`)
+
+  await download({ ...args, url })
+
+  return {
+    fileName: url.split('/').pop(),
+    installPath: path.join(args.installPath, '/go-ipfs/')
```
Comment (alanshaw) [inline src/index.js]:
```suggestion
    installPath: path.join(args.installPath, 'go-ipfs')
```

File: src/index.js [hunk 6]
```diff
@@ -25,94 +25,112 @@ const path = require('path')
...
+async function ensureVersion ({ version, distUrl }) {
+  const res = await get(`${distUrl}/go-ipfs/versions`)
+  const versions = res.split('\n').filter(el => el !== '')
+
+  if (versions.indexOf(version) === -1) {
+    throw new Error(`Version '${version}' not available`)
+  }
+}
+
+async function getURL ({ version, platform, arch, distUrl }) {
```
Comment (alanshaw) [inline src/index.js]:
```suggestion
async function getDownloadURL ({ version, platform, arch, distUrl }) {
```

File: src/index.js (hunk 4 above)
Comment (hacdias) [reply src/index.js]: I just kept request to use with the `unpack` function.

Comment (alanshaw) [reply src/index.js]: Not sure I follow? I don't think `unpack` requires request does it? If you just need a stream you can pass it `res.body` from `node-fetch`. See https://www.npmjs.com/package/node-fetch#streams

```js
function download ({ installPath, url }) {
  const res = await fetch(url)
  if (!res.ok) throw new Error(`unexpected status ${res.status}`)
  return unpack({ url, installPath, stream: res.body })
}
```

...right?

Comment (hacdias) [reply src/index.js]: Forget forget 😆 Done!

**Comment by hacdias** [comment]
@alanshaw updated 😄

**Comment by alanshaw** [comment]
@hacdias just noticed - can you update the readme - I think it's referencing files that don't exist now.

**Comment by hacdias** [comment]
@alanshaw done and I agree. As far as I can tell, only @daviddias has permission to publish new releases here.

**Comment by hacdias** [comment]
@daviddias could you release '0.4.19-1', please? 😄

**Comment by olizilla** [comment]
There was a [`0.4.19-0` release](https://github.com/ipfs/npm-go-ipfs-dep/releases/tag/v0.4.19-0) a couple of days ago.

@hacdias can you confirm that release is working as expected?

**Pull request merged**